pkg/simple_cache/lib/simple_cache.dart

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:identity_codec/identity_codec.dart';
+import 'package:logging/logging.dart';
+import 'cache_provider.dart';
+import 'src/providers/inmemory.dart';
+import 'src/providers/redis.dart';
+
+final _logger = Logger('simple_cache');
+
+/// Cache for objects of type [T], wrapping a [CacheProvider] to provide a
+/// high-level interface.
+///
+/// Cache entries are accessed using the indexing operator `[]`, this returns a
+/// [Entry<T>] wrapper that can be used to get/set data cached at given key.
+///
+/// **Example**
+/// ```dart
+/// final Cache<List<int>> cache = Cache.inMemoryCache(4096);
+///
+/// // Write data to cache
+/// await cache['cached-zeros'].set([0, 0, 0, 0]);
+///
+/// // Read data from cache
+/// var r = await cache['cached-zeros'].get();
+/// expect(r, equals([0, 0, 0, 0]));
+/// ```
+///
+/// A [Cache] can be _fused_ with a [Codec] using [withCodec] to get a cache
+/// that stores a different kind of objects. It is also possible to create
+/// a chuild cache using [withPrefix], such that all entries in the child
+/// cache have a given prefix.
+abstract class Cache<T> {
+  /// Get [Entry] wrapping data cached at [key].
+  Entry<T> operator [](String key);
+
+  /// Get a [Cache] wrapping of this cache with given [prefix].
+  Cache<T> withPrefix(String prefix);
```

Comment:
I'm wondering: it seems that we'll use `withPrefix`, `withTtl`, `withCodec` together most of the time. Maybe a single method of `fork({prefix, ttl, codec, logger})` would make this simpler?

Reply:
Maybe... We could add a `wrap({prefix, ttl, codec, logger})` or `child({...`, but let's see what the usage patterns are like first... we can easily refactor this later...